### PR TITLE
MLflow refresh for 2.0

### PR DESCRIPTION
**Pull request by santiagxf**
# Description

This PR includes changes for Mlflow 2.0 and a general refresh of samples.

# Checklist

- [x] I have read the [contribution guidelines](https://github.com/Azure/azureml-examples/blob/main/CONTRIBUTING.md)
- [x] Pull request includes test coverage for the included changes.

### CODE REVIEW

File: sdk/python/using-mlflow/deploy/mlflow_sdk_online_endpoints_progresive.ipynb [hunk 1]
```diff
@@ -0,0 +1,794 @@
+{
+ "cells": [
+  {
```
Comment (singankit) [inline sdk/python/using-mlflow/deploy/mlflow_sdk_online_endpoints_progresive.ipynb:3]: Overall this an amazing notebook. One thing I noticed is that notebook mentions to use AML SDK to get endpoint auth key. Otherwise rest of the notebook doesn't depend on AML SDK. Is there a way this can be achieved using mlflow sdk ? or by using mlflow sdk for invoking the endpoint rather than curl ?

File: sdk/python/using-mlflow/train-and-log/xgboost_classification_mlflow.ipynb [hunk 2]
```diff
@@ -0,0 +1,701 @@
...
+   "outputs": [],
+   "source": [
+    "pd.DataFrame(data=[run.data.metrics], index=[\"Value\"]).T"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    ">Pay attention how metrics calculated with Scikit-Learn where automatically tracked for us. None of them were manually added to the run. Also, MLflow uses naming conventions including the variable's names to help undestand what was logged. `X_test` was added to the name of the metric meaning that they correspond to the metric in the testing split of the dataset."
```
Comment (wamartin-aml) [inline sdk/python/using-mlflow/train-and-log/xgboost_classification_mlflow.ipynb:325]: "Pay attention to how...were automatically"
they correspond > it corresponds

File: sdk/python/using-mlflow/train-and-log/xgboost_classification_mlflow.ipynb [hunk 3]
```diff
@@ -0,0 +1,701 @@
...
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Let's explore artifacts that got logged in the run. This requires to use the MLflow client:"
```
Comment (wamartin-aml) [inline sdk/python/using-mlflow/train-and-log/xgboost_classification_mlflow.ipynb:332]: requires _us_ to use

File: sdk/python/using-mlflow/train-and-log/xgboost_classification_mlflow.ipynb [hunk 4]
```diff
@@ -0,0 +1,701 @@
...
+   "outputs": [],
+   "source": [
+    "classifier = mlflow.xgboost.load_model(f\"runs:/{run.info.run_id}/model\")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "See that the type returned by this method is an XGBoost model's classifier"
```
Comment (wamartin-aml) [inline sdk/python/using-mlflow/train-and-log/xgboost_classification_mlflow.ipynb:413]: remove "model's"

File: sdk/python/using-mlflow/train-and-log/xgboost_classification_mlflow.ipynb [hunk 5]
```diff
@@ -0,0 +1,701 @@
...
+   "outputs": [],
+   "source": [
+    "type(classifier)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "You can get prediction back from the model"
```
Comment (wamartin-aml) [inline sdk/python/using-mlflow/train-and-log/xgboost_classification_mlflow.ipynb:429]: get _the_ prediction

File: sdk/python/using-mlflow/train-and-log/xgboost_classification_mlflow.ipynb [hunk 6]
```diff
@@ -0,0 +1,701 @@
...
+   "source": [
+    "import numpy as np\n",
+    "from sklearn.preprocessing import OrdinalEncoder"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "We only need to transform the column `thal`. A convenient way to do this is by applying a `ColumnTransformer` to that column, the remaining columns will be sent directly to the model:"
```
Comment (wamartin-aml) [inline sdk/python/using-mlflow/train-and-log/xgboost_classification_mlflow.ipynb:503]: column, the remaining > column so that the remaining

File: sdk/python/using-mlflow/train-and-log/xgboost_classification_mlflow.ipynb [hunk 7]
```diff
@@ -0,0 +1,701 @@
...
+   "outputs": [],
+   "source": [
+    "pipeline"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "We can log this model in our run. Since this is a Scikit-Learn object, we will log it using such flavor instead of `xgboost`. Let's create a new complete run so we can see the difference."
```
Comment (wamartin-aml) [inline sdk/python/using-mlflow/train-and-log/xgboost_classification_mlflow.ipynb:565]: such > that

File: sdk/python/using-mlflow/train-and-log/xgboost_classification_mlflow.ipynb [hunk 8]
```diff
@@ -0,0 +1,701 @@
...
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Signatures\n",
+    "\n",
+    "One extra thing we need to take care of is the expected signature of the model. Signatures are use by MLflow to know what type of inputs are expected for a given model. This allows the model builder to be explicit about which types are being expected. In the first model we logged, all inputs needed to be numeric, including the column `thal`. However, our new pipeline can encode this values automatically so we can take `thal` values in string format."
```
Comment (wamartin-aml) [inline sdk/python/using-mlflow/train-and-log/xgboost_classification_mlflow.ipynb:574]: use > used
this > these

File: sdk/python/using-mlflow/train-and-log/xgboost_classification_mlflow.ipynb [hunk 9]
```diff
@@ -0,0 +1,701 @@
...
+    "with mlflow.start_run() as run:\n",
+    "    pipeline.fit(X_train, y_train)\n",
+    "    mlflow.sklearn.log_model(pipeline, artifact_path=\"pipeline\", signature=signature)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "> **Note:** You are not required to log the pipeline model manually as you can also turn `mlflow.sklearn.autolog()` on. If you do that, the model will automatically be logged for for by Scikit-Learn integration with MLflow. However, we have preferred to it in this way to show the different approaches and to be explit about logging a pipeline."
```
Comment (wamartin-aml) [inline sdk/python/using-mlflow/train-and-log/xgboost_classification_mlflow.ipynb:628]: for for > for you
explit > explicit

File: sdk/python/using-mlflow/train-and-log/xgboost_classification_mlflow.ipynb [hunk 10]
```diff
@@ -0,0 +1,701 @@
...
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "If we explore try to get this model back now:"
```
Comment (wamartin-aml) [inline sdk/python/using-mlflow/train-and-log/xgboost_classification_mlflow.ipynb:635]: "Get the model back for exploration:"

File: sdk/python/using-mlflow/train-and-log/xgboost_service_principal.ipynb [hunk 11]
```diff
@@ -0,0 +1,523 @@
...
+   "outputs": [],
+   "source": [
+    "df[\"thal\"] = df[\"thal\"].astype(\"category\").cat.codes"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "The encoded values looks then as follows:"
```
Comment (wamartin-aml) [inline sdk/python/using-mlflow/train-and-log/xgboost_service_principal.ipynb:217]: values looks then as > values look as

File: sdk/python/using-mlflow/train-and-log/xgboost_service_principal.ipynb [hunk 12]
```diff
@@ -0,0 +1,523 @@
...
+   "outputs": [],
+   "source": [
+    "df[\"thal\"].unique()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Let's split our dataset in train and test, so we can assess the performance of the model without overfitting the dataset."
```
Comment (wamartin-aml) [inline sdk/python/using-mlflow/train-and-log/xgboost_service_principal.ipynb:233]: in > into
test, > test

File: sdk/python/using-mlflow/train-and-log/xgboost_service_principal.ipynb [hunk 13]
```diff
@@ -0,0 +1,523 @@
...
+    "from xgboost import XGBClassifier\n",
+    "\n",
+    "model = XGBClassifier(use_label_encoder=False, eval_metric=\"logloss\")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "As soon as the `train` method is executed, MLflow will stat a run in Azure ML to start tracking the experiment's run. However, it is always a good idea to start the run manually so you have the run ID at hand quickly. This is not required though.\n",
```
Comment (wamartin-aml) [inline sdk/python/using-mlflow/train-and-log/xgboost_service_principal.ipynb:294]: stat > start
required > required,

File: sdk/python/using-mlflow/train-and-log/xgboost_service_principal.ipynb [hunk 14]
```diff
@@ -0,0 +1,523 @@
...
+    "model.fit(X_train, y_train, eval_set=[(X_test, y_test)], verbose=False)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Logging extra metrics\n",
+    "\n",
+    "Autolog capabilities in XGBoost will log metrics like validation loss, however, it won't log any specific metric in a classification problem. In this case, we are going to pay closer attention to our ability to detect heart condition while avoiding a type II error as much as possible. To calculate the metric, we are going to use our test dataset:"
```
Comment (wamartin-aml) [inline sdk/python/using-mlflow/train-and-log/xgboost_service_principal.ipynb:323]: same as comment for this same cell in other classification notebook

File: sdk/python/using-mlflow/train-and-log/xgboost_service_principal.ipynb [hunk 15]
```diff
@@ -0,0 +1,523 @@
...
+   "outputs": [],
+   "source": [
+    "pd.DataFrame(data=[run.data.params], index=[\"Value\"]).T"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Let's explore the metrics values:"
```
Comment (wamartin-aml) [inline sdk/python/using-mlflow/train-and-log/xgboost_service_principal.ipynb:411]: "metric values OR metrics' values"

File: sdk/python/using-mlflow/train-and-log/xgboost_service_principal.ipynb [hunk 16]
```diff
@@ -0,0 +1,523 @@
...
+   "outputs": [],
+   "source": [
+    "pd.DataFrame(data=[run.data.metrics], index=[\"Value\"]).T"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    ">Pay attention how metrics calculated with Scikit-Learn where automatically tracked for us. None of them were manually added to the run. Also, MLflow uses naming conventions including the variable's names to help undestand what was logged. `X_test` was added to the name of the metric meaning that they correspond to the metric in the testing split of the dataset."
```
Comment (wamartin-aml) [inline sdk/python/using-mlflow/train-and-log/xgboost_service_principal.ipynb:427]: "to how"
where > were
they correpond > it corresponds

File: sdk/python/using-mlflow/train-and-log/xgboost_service_principal.ipynb [hunk 17]
```diff
@@ -0,0 +1,523 @@
...
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Let's explore artifacts that got logged in the run. This requires to use the MLflow client:"
```
Comment (wamartin-aml) [inline sdk/python/using-mlflow/train-and-log/xgboost_service_principal.ipynb:434]: requires to > requires us to

File: sdk/python/using-mlflow/using-rest-api/using_mlflow_rest_api.ipynb [hunk 18]
```diff
@@ -117,13 +117,14 @@
...
   {
+   "attachments": {},
    "cell_type": "markdown",
    "id": "5ac77db1-5fb2-408a-b231-409820099494",
    "metadata": {},
    "source": [
     "## Querying the MLflow API\n",
     "\n",
-    "You can get the MLflow API from the Azure portal, or you can construct it manually as follows:"
+    "You can get the MLflow API from the Azure portal, or using the MLflow SDK as follows:"
```
Comment (wamartin-aml) [inline sdk/python/using-mlflow/using-rest-api/using_mlflow_rest_api.ipynb:127]: portal, > portal

File: sdk/python/using-mlflow/deploy/mlflow_sdk_online_endpoints_progresive.ipynb (hunk 1 above)
Comment (santiagxf) [reply sdk/python/using-mlflow/deploy/mlflow_sdk_online_endpoints_progresive.ipynb:3]: Indeed. I couldn't find a way to retrieve the authentication key. I don't know if it makes sense to return them as part of the endpoint properties?

File: sdk/python/using-mlflow/model-management/model_management.ipynb [hunk 19]
```diff
@@ -280,10 +301,13 @@
...
   },
   {
+   "attachments": {},
    "cell_type": "markdown",
    "metadata": {},
    "source": [
-    "You can query all the registered models in the registry using the MLflow client with the method `list_registered_models`."
+    "You can query all the registered models in the registry using the MLflow client with the method `search_registered_models`.\n",
+    "\n",
+    "> __MLflow 2.0 advisory__: In older versions of Mlflow (<2.0), please use `list_registered_models` instead."
```
[the remark this answers is not in the document]
Comment (dhmanigit) [reply sdk/python/using-mlflow/model-management/model_management.ipynb:310]: re-word to "search_registered_models is recommended instead of list_registered_models" ?

https://pyup.io/packages/pypi/mlflow/changelog
MLflow 1.10.0 - Added paginated `MlflowClient.search_registered_models` API (2939, 3023, 3027 ankitmathur-db; 2966, mparkhe)

File: sdk/python/using-mlflow/deploy/track_with_databricks_deploy_aml.ipynb [hunk 20]
```diff
@@ -0,0 +1,1008 @@
...
+   "metadata": {
+    "application/vnd.databricks.v1+cell": {
+     "inputWidgets": {},
+     "nuid": "08921861-db40-4ad7-8ecf-40792023a9ee",
+     "showTitle": false,
+     "title": ""
+    }
+   },
+   "source": [
+    "# Training models in Azure Databricks and deploying them on Azure ML\n",
```
Comment (dhmanigit) [inline sdk/python/using-mlflow/deploy/track_with_databricks_deploy_aml.ipynb:14]: A general comment to all notebooks: Better to include reference to official documentation for further clarifications or references. Will also help in cases where there are known gaps - for example, below link has a warning "Dual-tracking in a private link enabled Azure Machine Learning workspace is not supported by the moment. Configure exclusive tracking with your Azure Machine Learning workspace instead."

https://learn.microsoft.com/en-us/azure/machine-learning/how-to-use-mlflow-azure-databricks?tabs=azuremlsdk#dual-tracking-on-azure-databricks-and-azure-machine-learning

**Review by dhmanigit — APPROVED** [review]

File: sdk/python/using-mlflow/deploy/mlflow_sdk_online_endpoints_progresive.ipynb [hunk 21]
```diff
@@ -452,15 +494,91 @@
...
+   ]
+  },
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "#### 4.2.1 Create a sample request\n",
+    "\n",
+    "Azure Machine Learning requires the key `input_data` to be added to the input examples that you want to provide to the service. Notice that this is not the case of the command `mlflow model serve`.\n",
```
Comment (wamartin-aml) [inline sdk/python/using-mlflow/deploy/mlflow_sdk_online_endpoints_progresive.ipynb:579]: of the command > for the command

**Review by wamartin-aml — APPROVED** [review]

File: sdk/python/using-mlflow/deploy/track_with_databricks_deploy_aml.ipynb (hunk 20 above)
Comment (santiagxf) [reply sdk/python/using-mlflow/deploy/track_with_databricks_deploy_aml.ipynb:14]: Good point. I will make those updates!

File: sdk/python/using-mlflow/deploy/mlflow_sdk_online_endpoints_progresive.ipynb (hunk 1 above)
Comment (singankit) [reply sdk/python/using-mlflow/deploy/mlflow_sdk_online_endpoints_progresive.ipynb:3]: deployment_client.predict should take care of it internally depending the auth used for enpoint. Would that be an option?

**Review by singankit — APPROVED** [review]

File: sdk/python/using-mlflow/deploy/mlflow_sdk_online_endpoints_progresive.ipynb (hunk 1 above)
Comment (santiagxf) [reply sdk/python/using-mlflow/deploy/mlflow_sdk_online_endpoints_progresive.ipynb:3]: Added!
